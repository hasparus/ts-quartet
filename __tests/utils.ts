import { getDescription } from "./getDescription";

export function snapshot(validator: any) {
  const description = getDescription(validator);

  expect(description).toMatchSnapshot();
}
export function getExplanation(
  validator: { (arg0: any): void; explanations: any },
  value: any
) {
  validator(value);
  return validator.explanations;
}
export function isValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (!valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(true);
}
export function isNotValid(validator: (arg0: any) => any, value: any) {
  const valid = validator(value);
  if (valid) {
    console.log(JSON.stringify(value, null, 2));
  }
  expect(valid).toBe(false);
}

export function tables(validator: any, valids: any[], invalids: any[]) {
  for (const valid of valids) {
    isValid(validator, valid);
  }
  for (const invalid of invalids) {
    isNotValid(validator, invalid);
  }
}
