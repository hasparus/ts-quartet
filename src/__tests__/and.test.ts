import { v } from "../index";
import { snapshot, tables } from "./utils";

describe("v.compileAnd", () => {
  test("v.compileAnd()", () => {
    const validator = v.compileAnd();
    snapshot(validator);
  });
  test("v.compileAnd(1)", () => {
    const validator = v.compileAnd(1);
    snapshot(validator);
    tables(validator, [1], [2]);
  });
  test("v.compileAnd(1, 1)", () => {
    const validator = v.compileAnd(1, 1);
    snapshot(validator);
    tables(validator, [1], [2]);
  });
  test("v.compileAnd(1, 1, 2)", () => {
    const validator = v.compileAnd(1, 1, 2);
    tables(validator, [], [1, 2]);
    snapshot(validator);
  });
  test("v.compileAnd(v.positive, v.number)", () => {
    const validator = v.compileAnd(v.positive, v.number);
    tables(validator, [1, 0.1, Infinity, 10], [0, -1, -Infinity, "text"]);
    snapshot(validator);
  });
  test("v.compileAnd({ a: v.number }, { b: v.string})", () => {
    const validator = v.compileAnd({ a: v.number }, { b: v.string });
    tables(validator, [{ a: 1, b: "2" }], [{ a: "1", b: "2" }, { a: 1, b: 2 }]);
    snapshot(validator);
  });
});
