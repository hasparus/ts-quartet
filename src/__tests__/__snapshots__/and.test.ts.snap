// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v.compileAnd v.compileAnd() 1`] = `
Object {
  "_": "function () { return true; }",
  "explanations": Array [],
}
`;

exports[`v.compileAnd v.compileAnd({ a: v.number }, { b: v.string}) 1`] = `
Object {
  "0-0": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (typeof value.a !== 'number') return false
    return true
  }",
  "0-0.__validValues": Object {},
  "0-0.explanations": Array [],
  "1-0": "function validator(value) {
    validator.explanations = []
    if (!value) return false
    if (typeof value.b !== 'string') return false
    return true
  }",
  "1-0.__validValues": Object {},
  "1-0.explanations": Array [],
  "_": "function validator(value) {
    validator.explanations = []
    if (!validator['0-0'](value)) {
      validator.explanations.push(...validator['0-0'].explanations)
      return false
    }
    if (!validator['1-0'](value)) {
      validator.explanations.push(...validator['1-0'].explanations)
      return false
    }
    return true
  }",
  "explanations": Array [],
}
`;

exports[`v.compileAnd v.compileAnd(1) 1`] = `
Object {
  "_": "function (value) { return value === c; }",
  "explanations": Array [],
}
`;

exports[`v.compileAnd v.compileAnd(1, 1) 1`] = `
Object {
  "_": "function (value) { return value === c; }",
  "explanations": Array [],
}
`;

exports[`v.compileAnd v.compileAnd(1, 1, 2) 1`] = `
Object {
  "_": "function () { return false; }",
  "explanations": Array [],
}
`;

exports[`v.compileAnd v.compileAnd(v.positive, v.number) 1`] = `
Object {
  "_": "function validator(value) {
    validator.explanations = []
    if (value <= 0) return false
    if (typeof value !== 'number') return false
    return true
  }",
  "explanations": Array [],
}
`;
